### PR TITLE
Feature: Favorite Power Stones

### PR DESCRIPTION
**Pull request by saga-00**
## What
Highlights your favorite power stones. You can add/remove them with Shift Click.

<details>
<summary>Images</summary>

image

</details>

## Changelog New Features
+ Added Favorite Power Stones. - saga
    * Highlighted in the Thaumaturgy inventory.
    * Shift-click to add/remove them.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java [hunk 1]
```diff
@@ -212,6 +212,13 @@ public String toString() {
     @FeatureToggle
     public boolean powerStoneGuide = true;
 
+    @Expose
+    @ConfigOption(name = "Favorite Power Stone", desc = "Shows your favorite power stones. You can add/remove them by shift clicking a Power Stone,")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean favoritePowerStone = false;
+
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/InventoryConfig.java]:
```suggestion
    @ConfigOption(name = "Favorite Power Stone", desc = "Shows your favorite power stones. You can add/remove them by shift clicking a Power Stone.")
    @ConfigEditorBoolean
    @FeatureToggle
    public boolean favoritePowerStone = false;
```

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package at.hannibal2.skyhanni.features.inventory
+
+import at.hannibal2.skyhanni.SkyHanniMod
+import at.hannibal2.skyhanni.data.MaxwellAPI
+import at.hannibal2.skyhanni.data.MaxwellAPI.favoritePowers
+import at.hannibal2.skyhanni.data.MaxwellAPI.isThaumaturgyInventory
```
Comment (CalMWolfs) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt]: These shouldnt be statically imported

**Review by CalMWolfs — CHANGES_REQUESTED** [review]

**Review by CalMWolfs — APPROVED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    private val config get() = SkyHanniMod.feature.inventory
+    private val storage get() = ProfileStorageData.profileSpecific
+
+    private var highlightedSlots = mutableSetOf<Int>()
+
+    @SubscribeEvent
+    fun onBackgroundDrawn(event: GuiContainerEvent.BackgroundDrawnEvent) {
+        if (!isEnabled()) return
+        if (!MaxwellAPI.isThaumaturgyInventory(InventoryUtils.openInventoryName())) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt]: out of scope of the pr, but we should not expose regex check functions outside, rather cache the value in isInventory boolean. no need to test this regex each frame

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+        event.cancel()
+    }
+
+    @SubscribeEvent
+    fun onInventoryUpdated(event: InventoryUpdatedEvent) {
+        if (!isEnabled()) return
+        if (!MaxwellAPI.isThaumaturgyInventory(event.inventoryName)) return
+
+        event.inventoryItems.forEach { (slot, item) ->
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt]: We do not clear the highlighted slots list before adding into the list.

Therefore this breaks when switching to second page
grafik

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    fun onSlotClick(event: GuiContainerEvent.SlotClickEvent) {
+        if (!isEnabled() || !KeyboardManager.isShiftKeyDown()) return
+        if (!MaxwellAPI.isThaumaturgyInventory(InventoryUtils.openInventoryName())) return
+
+        val displayName = event.item?.name?.removeColor()?.trim() ?: return
+        val power = MaxwellAPI.getPowerByNameOrNull(displayName) ?: return
+
+        if (power in MaxwellAPI.favoritePowers) {
+            MaxwellAPI.favoritePowers -= power
+            highlightedSlots -= event.slotId
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt:44]: Please either use mutable val, or immutable var. but using both concepts at the same time is a bad design.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
Works fine in game. Just several smaller code cleanup ideas, and one minor bug.

File: src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt (hunk 3 above)
Comment (saga-00) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/FavoritePowerStone.kt]: I had done it before, but couldn't figure a way to fix a bug caused by it. If you can, please test what happens when you open the menu, go to tunings and come back to the main menu (without reopening it). The render used to not work in that situation, I don't know if it'll happen again.

**Review by hannibal002 — APPROVED** [review]
works for me

**Pull request merged**